File: tobas_setup_assistant/src/tobas_setup_assistant/parameter_getters/double_range.py
from overrides import override
from typing import Tuple, Optional
from PyQt5.QtCore import pyqtSignal, pyqtSlot
from PyQt5.QtWidgets import QHBoxLayout

from .base import ParamGetterWidget
from .utils import FloatGetter


class ParamGetterWidget_DoubleRange(ParamGetterWidget):
    value_changed = pyqtSignal(float, float)

    def __init__(
        self,
        param_name: str,
        description_text: Optional[str] = None,
        decimals: int = 3,
        minimum: float = -1e9,
        maximum: float = +1e9,
        single_step: float = 1.0,
        default: Tuple[float, float] = (0.0, 0.0),
        suffix: str = "",
    ) -> None:
        super().__init__(param_name, description_text)

        cols = QHBoxLayout()
        self._rows.addLayout(cols)

        self._min = FloatGetter("min", decimals, minimum, maximum, single_step, default[0], suffix)
        cols.addWidget(self._min)

        self._max = FloatGetter("max", decimals, minimum, maximum, single_step, default[1], suffix)
        cols.addWidget(self._max)

        self._min.value_changed.connect(self._on_value_changed)
        self._max.value_changed.connect(self._on_value_changed)

    @override
    def get(self) -> Tuple[float, float]:
        return self.min(), self.max()

    @override
    def set(self, src: Tuple[float, float]) -> None:
        self._min.set(src[0])
        self._max.set(src[1])

    def min(self) -> float:
        return self._min.get()

    def max(self) -> float:
        return self._max.get()

    def is_valid(self) -> bool:
        return self.min() <= self.max()

    @pyqtSlot(float)
    def _on_value_changed(self, _: float) -> None:
        self.value_changed.emit(self.min(), self.max())

File: tobas_gui/tobas_setup_assistant/src/param_getters/double_range.cpp
#include "tobas_setup_assistant/param_getters/double_range.hpp"

namespace gui
{
namespace sa
{
ParamGetterWidget_DoubleRange::ParamGetterWidget_DoubleRange(const QString& param_name, const QString& description_text)
  : super(param_name, description_text)
{
  const auto cols = new QHBoxLayout();
  rows_->addLayout(cols);

  min_ = new DoubleGetter("min");
  cols->addWidget(min_);

  max_ = new DoubleGetter("max");
  cols->addWidget(max_);

  connect(min_, &DoubleGetter::valueChanged, this, &self::onValueChanged);
  connect(max_, &DoubleGetter::valueChanged, this, &self::onValueChanged);
}

std::pair<double, double> ParamGetterWidget_DoubleRange::getValue() const
{
  return { min(), max() };
}

bool ParamGetterWidget_DoubleRange::setValue(const std::pair<double, double>& src)
{
  min_->setValue(src.first);
  max_->setValue(src.second);
  return true;
}

void ParamGetterWidget_DoubleRange::setDecimals(int decimals)
{
  min_->setDecimals(decimals);
  max_->setDecimals(decimals);
}

void ParamGetterWidget_DoubleRange::setMinimum(double minimum)
{
  min_->setMinimum(minimum);
  max_->setMinimum(minimum);
}

void ParamGetterWidget_DoubleRange::setMaximum(double maximum)
{
  min_->setMaximum(maximum);
  max_->setMaximum(maximum);
}

void ParamGetterWidget_DoubleRange::setSingleStep(double single_step)
{
  min_->setSingleStep(single_step);
  max_->setSingleStep(single_step);
}

void ParamGetterWidget_DoubleRange::setSuffix(const QString& suffix)
{
  min_->setSuffix(suffix);
  max_->setSuffix(suffix);
}

double ParamGetterWidget_DoubleRange::min() const
{
  return min_->getValue();
}

double ParamGetterWidget_DoubleRange::max() const
{
  return max_->getValue();
}

bool ParamGetterWidget_DoubleRange::isValid() const
{
  return min() <= max();
}

void ParamGetterWidget_DoubleRange::onValueChanged(double)
{
  Q_EMIT valueChanged({ min(), max() });
}
}  // namespace sa
}  // namespace gui
